python/cudf/cudf/core/dataframe.py

```diff
@@ -1453,84 +1420,91 @@ def _get_columns_by_label(self, labels, downcast=False):
-                    result[col] = fallback(rhs[col], _reverse_op[fn])
-        elif isinstance(other, Series):
-            other_cols = other.to_pandas().to_dict()
-            df_cols = self._column_names
-            result_cols = df_cols + tuple(
-                col for col in other_cols if col not in df_cols
+            operands = {
+                name: (
+                    lcol,
+                    rhs._data[name]
+                    if name in rhs._data
+                    else (fill_value or None),
+                    reflect,
+                    fill_value if name in rhs._data else None,
+                )
+                for name, lcol in lhs._data.items()
+            }
+            for name, col in rhs._data.items():
+                if name not in lhs._data:
+                    operands[name] = (
+                        col,
+                        (fill_value or None),
+                        not reflect,
+                        None,
+                    )
+        elif isinstance(rhs, Series):
+            # Note: This logic will need updating if any of the user-facing
+            # binop methods (e.g. DataFrame.add) ever support axis=0/rows.
+            right_dict = dict(zip(rhs.index.values_host, rhs.values_host))
+            left_cols = lhs._column_names
+            # mypy thinks lhs._column_names is a List rather than a Tuple, so
+            # we have to ignore the type check.
+            result_cols = left_cols + tuple(  # type: ignore
+                col for col in right_dict if col not in left_cols
             )
-
+            operands = {}
             for col in result_cols:
-                l_opr = (
-                    self[col]
-                    if col in df_cols
-                    else Series(as_column(np.nan, length=len(self)))
-                )
-                r_opr = other_cols.get(col)
-                result[col] = op(l_opr, r_opr)
-
-        elif isinstance(other, (numbers.Number, cudf.Scalar)) or (
-            isinstance(other, np.ndarray) and other.ndim == 0
-        ):
-            for col in self._data:
-                result[col] = op(self[col], other)
+                if col in left_cols:
+                    left = lhs._data[col]
+                    right = right_dict[col] if col in right_dict else None
+                else:
+                    # We match pandas semantics here by performing binops
+                    # between a NaN (not NULL!) column and the actual values,
+                    # which results in nans, the pandas output.
+                    left = as_column(np.nan, length=lhs._num_rows)
+                    right = right_dict[col]
+                operands[col] = (left, right, reflect, fill_value)
         else:
-            raise NotImplementedError(
-                "DataFrame operations with " + str(type(other)) + " not "
-                "supported at this time."
-            )
-        return result
+            return NotImplemented
```

Comment:
Should we add a message for this exception?

Reply:
`NotImplemented` isn't an exception, [it's a special return value that triggers reflected binops](https://docs.python.org/3/library/constants.html#NotImplemented). The purpose is to ensure that binary operations work both ways. Here's an example: `cudf_series + pandas_series` triggers `cudf_series.__add__(pandas_series)`, so if a cudf `Series` knows how to add a pandas `Series` this will work. On the other hand, `pandas_series + cudf_series` will trigger `pandas_series.__add__(cudf_series)`, and `pandas` doesn't know about `cudf` so this will fail. However, if `pandas_series.__add__` does what it's supposed to and returns `NotImplemented` rather than just raising an exception, Python will detect this and automatically call `cudf_series.__radd__(pandas_series)` (the reflected binop). By implementing the same behavior in `cudf`, we allow packages extending binops to do the same (for example to support addition of cuspatial's `DataFrame` to `cudf.DataFrame` if that requires special logic for the geo columns. 

This is a long way of saying, there's no message to be added here, but I could add a comment if you think it would help future readers of the code who aren't familiar with Python's reflected ops.

Reply:
I see, TIL. Thanks for explaining this Vyas. Not action needed here.